### PR TITLE
Allow for subinterfaces

### PR DESCRIPTION
**Pull request by vanosg**
There's probably a better way to go about this, but...

Pipework will work well using sub-interfaces (eth0:1, eth0:2, etc) but when you run it as such, the current checking method does not detect it as a valid interface and quits. This patch will parse anything to the left of the colon in a sub-interface to get past the check. Everything else still works the same. My example code:

./pipework eth0:1 $(docker run -d -t -i --name=test --privileged=true ubunutu /sbin/my_init -- bash -l) 192.168.10.55/24

That would fail under the current code, but is allowed under this patch and creates a sub-interface eth0:1 on the host machine, along with eth1 of the same IP in the container. 

Like I said, probably a better way to do it, but this worked for me. Thanks for a great project!

### CODE REVIEW

**Comment by jpetazzo** [comment]
Sorry for the delay. I'm revisiting this, and I still think we disagree.

Can you point me to your version of pipework that has the appropriate fix, so that I can check how it behaves?

Thank you.

**Comment by vanosg** [comment]
Sure - the change is actually submitted as part of this pull request - https://github.com/jpetazzo/pipework/pull/86/commits 

It's just a simple one line mod to allow the use of a subinterface, should the user choose to use one.

**Comment by jpetazzo** [comment]
OK! I ran a few tests.

If I run this:
    pipework eth0:1 $(docker run -d nginx) 192.168.10.71/24

Then I can `curl 192.168.10.71` (from another machine).

But if I do this:
    pipework eth0 $(docker run -d nginx) 192.168.10.71/24

It works exactly the same way (i.e. I can still `curl` from another machine).

Do you see the same behavior on your setup?

**Comment by vanosg** [comment]
I am unable to test right now, but they should both work the same way (and remember them as working the same way previously). For those of us that use sub-interfaces, this patch will allow that. I think we’ve shown that there are larger-reaching impacts by doing this. 

 

If you recall, the original problem I had was the code actively denied me from using the sub-interfaces. The existing code limits interfaces to those listed in /sys/class/net – there are other interfaces that can exist (and work the same way as those in /sys/class/net) . This patch will allow the user to use pipework in conjunction with those interfaces, which a user cannot use as the code exists currently.

**Comment by jpetazzo** [comment]
Hi,

I just did another round of testing. If I run:

- `pipework eth0:1 ...` (with your version)
- or `pipework eth0 ...` (with my version)

... the end result is exactly the same. If you're still interested into merging this, let me know if it works differently on your machine.

Thank you!

**Comment by vanosg** [comment]
I’ve since moved on from this project. I’m not sure you understand the fundamental reason behind this- subinterfaces (eth0:1, eth1:5, etc) are a way of adding additional IPs to the host machine. They are used in many scenarios. I had to use them for my last project. Your current codebase does not allow sub-interfaces to be used because of how you collect the interface information. “/sys/class/net/” only lists physical interfaces, not sub-interfaces. 

 

If someone has sub-interfaces and wants to use them, the codebase as it is does not allow for them. This patch enables sub-interfaces to be used, if a user has them on the machine. 

 

I don’t want to get distracted with all the conversation on what they do, what purpose the serve, etc etc, but rather focus on the fact that some people use them, and if they do, pipework can’t support it ONLY because of the directory being checked not listing sub-interfaces. Literally, this check in the code is the only thing preventing them from running. As you just proved, they work fine in the code- it is just this check that prevents them from being bound to only because they’re not listed in this directory. 

 

Bottom line- if someone is using eth0:1 right now with docker, they can’t use pipework on that interface. This patch allows it. I hope that helps, and hope you merge it, it can be a big help to large multiple networks being run through one box. 

 

Thanks for all the hard work!
